modin/pandas/test/test_groupby.py

```diff
@@ -1113,6 +1116,79 @@ def eval___getattr__(modin_groupby, pandas_groupby, item):
     )
 
 
+def eval___getitem__(md_grp, pd_grp, additional_tests=None):
+    if isinstance(md_grp, pd.groupby.SeriesGroupBy):
+        return
+
+    additional_tests = (
+        tuple()
+        if additional_tests is None
+        else additional_tests
+        if is_list_like(additional_tests)
+        else (additional_tests,)
+    )
+
+    def comparator(df1, df2):
+        if df1.ndim == 2:
+            df_equals(df1.sort_index(axis=1), df2.sort_index(axis=1))
```

Comment:
Why do we need to sort index?

Reply:
Because we can't ensure the same order as pandas. When we select columns in pandas we get the columns in the same **order as we selected**:
```python
>>> pandas_df = (
    a b c e f g h i j
    . . . . . . . . .
    . . . . . . . . .
)
>>> pandas_df.groupby("a")["e", "b", "g"].sum()
    e b g
    . . .
    . . .
```
When working with Modin the partitioning comes into play and it becomes difficult and expensive to ensure the order as Pandas, in the current implementation Modin preserves the **frame's** order of the columns, and that's why:
in the following example, placing selected columns in the `[e b g]` order would require of splitting the second column partition into two and placing the "e" partition before the "b" partition:
```python
>>> modin_df = (
    a b c | e f g | h i j
    . . . | . . . | . . .
    . . . | . . . | . . .
)

>>> modin_df.groupby("a")["e", "b", "g"].sum()
    e | b | g
    . | . | .
    . | . | .
```
the present implementation doesn't do that and preserves the partitioning:
```python
>>> modin_df.groupby("a")["e", "b", "g"].sum()
    b | e g |
    . | . . |
    . | . . |
```

Reply:
One of the functions that are guilty of not preserving the passed indices order is the `modin_frame._get_dict_of_blocks`, which is used to determine partition and its internal indices to apply aggregation function based on the passed external indices. `_get_dict_of_blocks` function sorts passed labels so their order is the same as the frame's and so returns the required indices in the frame's order.

We already have an issue for that #2552 (in #2552 it relates to non-preserving assignment order), this issue is harder than it looks because of the potential overheads of shuffling partition's data (described in the above comment), so we haven't found any elegant solution yet.

@gshimansky reindexing groupby result requires data shuffling along the row axis, which might slowdown selection-groupby in about 15-20%

<details><summary>Slowdown example</summary>

```python
import modin.pandas as pd
from modin.pandas.test.utils import test_data_values
from asv_bench.benchmarks.utils import execute

import timeit

md_df = pd.DataFrame(test_data_values[0])

by = md_df.columns[0]
selection = md_df.columns[[-1, 0, -2, 34, -34, 1, -3, 2]].tolist()
grp = md_df.groupby(by)[selection]

def no_reindex():
    res = grp.sum()
    execute(res)

def reindex():
    res = grp.sum()
    reindexed_res = res.reindex(columns=selection)
    execute(reindexed_res)

nrepeats = 100
no_reindex_time = timeit.timeit(no_reindex, number=nrepeats)
reindex_time = timeit.timeit(reindex, number=nrepeats)

print(no_reindex_time, reindex_time)

# Possible output:
#   13.990928617015015 16.142298899008892
```

</details>

Reply:
I would add some note, probably in function description or in release notes, that our behavior is different from Pandas, and how it is possible get the behavior identical to Pandas.

Reply:
Added this note to the `__getitem__`'s docstring:

<details><summary>Note</summary>

```
        Notes
        -----
        Unlike pandas, modin always selects columns in the same order as they're located
        in the source frame, discarding the order provided to the `__getitem__`. If the
        order is matters, you can reindex GroupBy result with the selection:
        ``result.reindex(columns=result.columns[:result.shape[1] - len(selection)].tolist() + selection)``
```

</details>

Reply:
Ok thank you. Just replace "is matters" with just "matters".